### PR TITLE
feat: Bionemo on Eks

### PR DESCRIPTION
**Pull request by dorukozturk**
### What does this PR do?

Adds a bionemo example blueprint for training a model.

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

This is the EKS implementation of the bionemo framework's ESM-1nv model training example. https://docs.nvidia.com/bionemo-framework/latest/notebooks/model_training_esm1nv.html

### Motivation

Some AWS customers are looking to try this blueprint.

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [x] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: ai-ml/bionemo/addons.tf [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    nvidia-device-plugin = {
+      description      = "A Helm chart for NVIDIA Device Plugin"
+      namespace        = "nvidia-device-plugin"
+      create_namespace = true
+      chart            = "nvidia-device-plugin"
+      chart_version    = "0.14.0"
+      repository       = "https://nvidia.github.io/k8s-device-plugin"
+      values           = [file("${path.module}/helm-values/nvidia-values.yaml")]
+    }
+  }
```
Comment (vara-bonthu) [inline ai-ml/bionemo/addons.tf:75]: remove this block and `nvidia-values.yaml` file and simply use the below line under "eks_data_addons" module

#---------------------------------------------------------------
# Data on EKS Kubernetes Addons
#---------------------------------------------------------------
module "eks_data_addons" {
  source  = "aws-ia/eks-data-addons/aws"
  version = "~> 1.30" # ensure to update this to the latest/desired version
 oidc_provider_arn = module.eks.oidc_provider_arn

`enable_nvidia_device_plugin = true`
}

File: ai-ml/bionemo/eks.tf [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    iam_role_additional_policies = {
+      # Not required, but used in the example to access the nodes to inspect mounted volumes
+      AmazonSSMManagedInstanceCore = "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore"
+    }
+  }
+
+  eks_managed_node_groups = {
+    #  We recommend to have a MNG to place your critical workloads and add-ons
+    #  Then rely on Karpenter to scale your workloads
+    #  You can also make uses on nodeSelector and Taints/tolerations to spread workloads on MNG or Karpenter provisioners
```
Comment (vara-bonthu) [inline ai-ml/bionemo/eks.tf:68]: Create a core node group for running system pods
e.g., https://github.com/awslabs/data-on-eks/blob/5f96cb3368f141edd919b09ba31476284d67c95b/ai-ml/emr-spark-rapids/eks.tf#L86

File: ai-ml/bionemo/eks.tf [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    gpu1 = {
+      name        = "gpu-node-grp"
+      description = "EKS Node Group to run GPU workloads"
+      # Filtering only Secondary CIDR private subnets starting with "100.".
+      # Subnet IDs where the nodes/node groups will be provisioned
+      subnet_ids = compact([for subnet_id, cidr_block in zipmap(module.vpc.private_subnets, module.vpc.private_subnets_cidr_blocks) :
+        substr(cidr_block, 0, 4) == "100." ? subnet_id : null]
+      )
+
+      ami_id     = "ami-074b2f0953ea3375c"
```
Comment (vara-bonthu) [inline ai-ml/bionemo/eks.tf:78]: Is there a reason that you are using custom AMI here? Did you not try and use the ` ami_type = "AL2_x86_64_GPU"`

File: ai-ml/bionemo/addons.tf [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  #---------------------------------------
+  enable_aws_cloudwatch_metrics = true
+  aws_cloudwatch_metrics = {
+    values = [templatefile("${path.module}/helm-values/aws-cloudwatch-metrics-values.yaml", {})]
+  }
+
+  #---------------------------------------
+  # Metrics Server
+  #---------------------------------------
+  enable_metrics_server = true
```
Comment (vara-bonthu) [inline ai-ml/bionemo/addons.tf:57]: remove this add-on if its not being used

File: ai-ml/bionemo/versions.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      source  = "hashicorp/http"
+      version = ">= 3.3"
+    }
+  }
+
+  # ##  Used for end-to-end testing on project; update to suit your needs
+  # backend "s3" {
+  #   bucket = "doeks-github-actions-e2e-test-state"
+  #   region = "us-west-2"
+  #   key    = "e2e/mlflow/terraform.tfstate"
```
Comment (vara-bonthu) [inline ai-ml/bionemo/versions.tf:35]: repalce mlflow with bionemo

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+```
+
+</CollapsibleContent>
+
+<CollapsibleContent header={<h3><span>Verify Deployment</span></h3>}>
+
+Update local kubeconfig so we can access kubernetes cluster
+
+```bash
+aws eks update-kubeconfig --name bionemo-cluster #or whatever you used for EKS cluster name
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:81]: name of the cluster is not matching with the varaibale name. change this to `bionemo-on-eks`

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+```
+
+Install Kubeflow Training Operator:
+```bash
+kubectl apply -k "github.com/kubeflow/training-operator/manifests/overlays/standalone?ref=v1.7.0"
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:86]: Could you add this to [EKS Data Addons repo](https://github.com/aws-ia/terraform-aws-eks-data-addons) and deploy it as an add-on? We want to reduce the number of manual commands in our blueprints. If they dont have a helm char then you can create a local helm chart as shown here https://github.com/aws-ia/terraform-aws-eks-data-addons/tree/main/helm-charts

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+```bash
+kubectl get pods -A
+```
+Make sure training-operator, nvidia-device-plugin and fsx-csi-controller pods are running and healthy.
+
+#### Run bionemo jobs
+
+Once we are sure everything is working, we can start submitting jobs to our clusters.
+
+First job we are going to run is uniref50-job which will download and split the data for efficient processing.
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:113]: Explain bit more about this job to the users. Tools use and frameworks are used in this job.

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+```bash
+cd training
+kubectl apply -f uniref50-job.yaml
+```
+
+If you look at the yaml manifest you would see it is a kubernetes Job, which simply runs a python script within the bionemo container.
+
+This process takes relatively a long time (~50-60 hours). In order to see if it is working check the logs from this pod:
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:122]: Add some details on why it takes longer. Is this a trianing job? How much data its processing? Is this job processing data from S3 or Fsx etc..

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+[NeMo I 2024-02-26 23:16:21 preprocess:255] Creating train split...
+```
+
+After this job is done, our processed dataset will be placed in /fsx/processed directory.
+
+After this step we can run the training job by running:
+
+```bash
+cd training
+kubectl apply -f esm1nv_pretrain-job.yaml
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:143]: What does this job do? How is this related to the previous job?

File: website/docs/blueprints/ai-ml/bionemo.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+```
+
+This manifest uses kubeflow's Pytorch training CRD. There are bunch of parameters within this manifest that could be modified. See [Bionemo's Docs](https://docs.nvidia.com/bionemo-framework/latest/notebooks/model_training_esm1nv.html) for detailed explanations of each parameter and how to tune them.
+
+According to Training Operator's documentation, ...worker-0 is the main process. In order to check the process:
```
Comment (vara-bonthu) [inline website/docs/blueprints/ai-ml/bionemo.md:148]: Write some conclusions for the document. Who can benefit from this blueprint? Also, provide some details about why you chose P3 instances and what P3 offers compared to other GPU instances.

**Review by vara-bonthu — COMMENTED** [review]
Thanks @dorukozturk for the new bionemo blueprint. I left some minor comments. I feel that we need more details to the website doc similar to writing any aws blog.

File: ai-ml/bionemo/eks.tf (hunk 3 above)
Comment (dorukozturk) [reply ai-ml/bionemo/eks.tf:78]: https://github.com/awslabs/amazon-eks-ami/issues/1697 seems like there are some issues so I decided to pin the ami

File: website/docs/blueprints/ai-ml/bionemo.md (hunk 7 above)
Comment (dorukozturk) [reply website/docs/blueprints/ai-ml/bionemo.md:86]: ok I was going to ask about this. I can do that

**Comment by dorukozturk** [comment]
@vara-bonthu Incorporated the changes, PTAL when you have time

File: ai-ml/bionemo/eks.tf (hunk 3 above)
Comment (bryantbiggs) [reply ai-ml/bionemo/eks.tf:110]: its better to use the release version on managed node groups so that it still treats the AMI as an EKS AMI - using an AMI ID will make it treat it as a custom AMI

You can use the release version to pin to a release to the GPU kernel module changes:
```hcl
  ami_type            = "AL2_x86_64_GPU"
  ami_release_version = "1.29.0-20240213"
  
  # Now you can remove this
  # enable_bootstrap_user_data = true
```

File: website/docs/gen-ai/training/bionemo.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+[NeMo I 2024-02-26 23:16:21 preprocess:255] Creating train split...
+```
+
+After finishing this task, the processed dataset will be saved in the `/fsx/processed` directory. Once this is done, we can move forward and start the `pre-training` job by running the following command:
+
+Following this, we can proceed to execute the pre-training job by running:
+
+```bash
+cd training
+kubectl apply -f esm1nv_pretrain-job.yaml
```
Comment (vara-bonthu) [inline website/docs/gen-ai/training/bionemo.md:156]: Show the output of `kubectl get pods` for this job. Explain how many nodes and pods are used by this job.

File: website/docs/gen-ai/training/bionemo.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+#### Step1: Initiate the Uniref50 Data Preparation Task:
+
+The first task, named the `uniref50-job.yaml`, involves downloading and partitioning the data to enhance processing efficiency. This task specifically retrieves the `uniref50 dataset` and organizes it within the FSx for Lustre Filesystem. This structured layout is designed for training, testing, and validation purposes. You can learn more about the uniref dataset [here](https://www.uniprot.org/help/uniref).
+
+To execute this job, navigate to the `examples\training` directory and deploy the `uniref50-job.yaml` manifest using the following commands:
+
+```bash
+cd training
+kubectl apply -f uniref50-job.yaml
```
Comment (vara-bonthu) [inline website/docs/gen-ai/training/bionemo.md:125]: Show the output of `kubectl get pods` for this job. Explain how many nodes and pods are used by this job.

File: ai-ml/bionemo/examples/training/esm1nv_pretrain-job.yaml [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      - type: Resource
+        resource:
+          name: cpu
+          target:
+            type: Utilization
+            averageUtilization: 80
+  nprocPerNode: "8"
+  pytorchReplicaSpecs:
+    Worker:
+      replicas: 16
```
Comment (vara-bonthu) [inline ai-ml/bionemo/examples/training/esm1nv_pretrain-job.yaml:21]: Does this creates 16 pods and each pod requests one GPU?

**Review by vara-bonthu — APPROVED** [review]
